### PR TITLE
Rewrite New-TemporaryFile in C# #2732

### PR DESCRIPTION
**Pull request by kittholland**
Literal translation of New-TemporaryFile to C# cmdlet.

### CODE REVIEW

**Comment by powercode** [comment]
And the command to the UNIX version of the psd1 to fix the build error.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                    ErrorRecord errorRecord = new ErrorRecord(
+                        e,
+                        "NewTemporaryFileWriteError",
+                        ErrorCategory.WriteError,
+                        System.Environment.GetEnvironmentVariable("TEMP")
+                    );
+                    WriteError(errorRecord);
+                    return;
+                }
+            FileInfo file = new FileInfo(filePath);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Please correct indentation.
`filePath` can be `null` here and raise exception.
Original cmdlet wrap this:
```C#
if (filePath)
{
   FileInfo file = new FileInfo(filePath);
   WriteObject(file); 
}
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+
+/********************************************************************++
+Copyright (c) Microsoft Corporation.  All rights reserved.
+--********************************************************************/
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: @PowerShellTeam Is it correct? What is a policy about copyright?

Comment (kittholland) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: First cmdlet I've writte and I saw that at the top of every other file, wasn't really sure what the protocol was. It's probably wrong. Removing now.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs (hunk 1 above)
Comment (kittholland) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Thanks, I wasn't sure what condition could fall into that, that would not trip the catch return. Thanks for clarifying.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+                    );
+                    WriteError(errorRecord);
+                    return;
+                }
+            FileInfo file = new FileInfo(filePath);
+            WriteObject(file);
+            }
+        }
+    }
+}
```
Comment (kittholland) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Done.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs (hunk 2 above)
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: @kittholland :-) My question was not addressed to you. Please don't rush to remove.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                    filePath = Path.GetTempFileName();                
+                }
+                catch (Exception e)
+                {
+                    WriteError(
+                        new ErrorRecord(
+                            e,
+                            "NewTemporaryFileWriteError",
+                            ErrorCategory.WriteError,
+                            System.Environment.GetEnvironmentVariable("TEMP")));
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Repeating code. Please use a variable.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+                    return;
+                }
+                //filePath may be null, only writeobject if there is a value
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: We usually use a space after `//`.
And the comment is obvious and likely unnecessary.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs (hunk 4 above)
Comment (kittholland) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Removed duplicate GetEnvironmentVariable call. Thanks

**Comment by iSazonov** [comment]
@kittholland Thanks for your contribution!

My question about the copyright header was not addressed to you. Please don't rush to remove it.
With this in mind, the PR is LGTM. Please wait review from PG.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    /// The implementation of the "New-TemporaryFile" cmdlet
+    /// </summary>    
+    [Cmdlet(VerbsCommon.New, "TemporaryFile", SupportsShouldProcess = true, HelpUri = "https://go.microsoft.com/fwlink/?LinkId=526726")]
+    [OutputType(typeof(System.IO.FileInfo))]
+    public class NewTemporaryFileCommand : Cmdlet
+    {      
+        /// <summary>
+        /// returns a TemporaryFile
+        /// </summary>
+        protected override void BeginProcessing()
```
Comment (lzybkr) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Putting the logic of a simple cmdlet in the `end` block does protect against stupid pipelines, e.g.

```powershell
function f { begin { write-host 'in begin f'}}
function g { end { write-host 'in end g'}}
g | f
```

This will print
```
in begin f
in end g
```
which is a little weird that the logic for a downstream cmdlet finishes before the upstream command (note no objects are being piped).

In others, you could more easily write some obfuscated (as in, hard to figure out the real logic) pipelines if you use `begin` instead of `end`.

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Thanks @lzybkr, this is a great example. So it looks the guideline should be using `EndProcessing` (or `End` block in script) for simple logic that doesn't require pipeline input.

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: I think the reason we have those cmdlets with main logic in `Begin`/`BeginProcessing` is that the pattern was copied from other existing cmdlets, without knowing why it should be put in `Begin`/`BeginProcessing`.

@kittholland could you please move the main logic to `EndProcessing` method given the discussion here? Thanks!

Comment (kittholland) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Yes I will move it.

I left as is because this is an existing cmdlet and did not want to potentially alter any existing implementations (and I did not feel that I could fully comprehend what those possible scenarios would have been).

Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: Yep, totally understood. Thanks!

Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs]: This implies that we need to open a new Issue with analysing what cmdlets must be corrected.

**Comment by JamesWTruher** [comment]
Are the tests in `test/powershell/Modules/Microsoft.PowerShell.Utility/NewTemporaryFile.Tests.ps1` sufficient, or should there be more?

**Comment by iSazonov** [comment]
It would be nice to test i/o on the file.

**Comment by kittholland** [comment]
I am happy to add an additional test regarding IO but I am uncertain if testing IO after the new file was generated is within the scope of this command.

I could validate additional attributes of the resulting file, for example that the extension is tmp, that the length is 0, and that it was created during the span of testing, but I'm not certain there are any conditions that would ever cause those tests to fail.

**Comment by iSazonov** [comment]
I believe that there is no need to examine the underlying level. We need to check only the cmdlet.
```powershell
$tmp=New-TemporaryFile
Get-Content -Path $tmp.FullName | Should Be $null
$teststring ="any test string"
$teststring | Out-File -Path $tmp.FullName
Get-Content -Path $tmp.FullName | Should Be $teststring
Remove-Item $tmp.FullName -Force
```

**Review by lzybkr — APPROVED** [review]

**Comment by lzybkr** [comment]
We can merge as soon as the conflicts are resolved.

**Comment by kittholland** [comment]
Conflicts have been resolved, thanks.

**Comment by lzybkr** [comment]
@kittholland - can you rebase? The PR now has a bunch of commits it shouldn't.

File: src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1 [hunk 7]
```diff
@@ -2,7 +2,7 @@
 GUID="1DA87E53-152B-403E-98DC-74D7B4D63D59"
 Author="Microsoft Corporation"
 CompanyName="Microsoft Corporation"
-Copyright="� Microsoft Corporation. All rights reserved."
+Copyright="� Microsoft Corporation. All rights reserved."
```
Comment (lzybkr) [inline src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1]: Revert this change please.

File: src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1 [hunk 8]
```diff
@@ -405,7 +374,7 @@ function Format-Hex
                     # 170..180 | format-hex
                     #           Path:
                     #           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
-                    #00000000   AA AB AC AD AE AF B0 B1 B2 B3 B4                 �����������
+                    #00000000   AA AB AC AD AE AF B0 B1 B2 B3 B4                 �����������
```
Comment (lzybkr) [inline src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1]: Revert this change please.

File: src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1 [hunk 9]
```diff
@@ -418,9 +387,9 @@ function Format-Hex
                     # 
                     #            00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F
                     # 
-                    # 00000000   AA 00 00 00 AB 00 00 00 AC 00 00 00 AD 00 00 00  �...�...�...�...
-                    # 00000010   AE 00 00 00 AF 00 00 00 B0 00 00 00 B1 00 00 00  �...�...�...�...
-                    # 00000020   B2 00 00 00 B3 00 00 00 B4 00 00 00              �...�...�...
+                    # 00000000   AA 00 00 00 AB 00 00 00 AC 00 00 00 AD 00 00 00  �...�...�...�...
```
Comment (lzybkr) [inline src/Modules/Shared/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psm1]: Revert this change please.

**Review by lzybkr — COMMENTED** [review]
Some unicode characters got messed up, please fix.

**Comment by iSazonov** [comment]
@kittholland Thanks for your contribution!
